### PR TITLE
Feature lumina image

### PR DESCRIPTION
**Pull request by kohya-ss**
- [x] check basic SD3/FLUX/SDXL/SD training
- [x] system prompt handling
- [x] cache text encoder outputs without disk 
- [x] minimal inference script

### CODE REVIEW

**Comment by kohya-ss** [comment]
By default, `system_prompt` is now added by tokenizer. Since the prompts during training is positive, it should be fine to add system prompts for them.

For the negative prompt during sample generation, the system prompt is not added with a flag for `tokenize`.
Training should probably work without any problems. 

I also added workarounds for `gated repo` issue for Gemma2 in pytest on Github Actions.

@sdbds @rockerBOO 
I would appreciate it if you could check it out.

A minimal inference script will be added.

**Comment by GumGum10** [comment]
Hi, really appreciate the work that you are doing on this! Me and a few other people have tried using this to train a LoRA on the new Neta finetune of Lumina2 (https://huggingface.co/neta-art/NetaLumina_Alpha).
We've all experienced the same issue, after training for a few steps (usually ~90), the LoRA seems to explode. The inference images come up with a lot of noise, like the model is collapsing.
Example:
image

Config:
```pretrained_model_name_or_path = "E:/Apps/ComfyUI_windows_portable/ComfyUI/output/diffusion_models/neta_train.safetensors"
gemma2 = "E:/Apps/trainLumi/models/gemma-2-2b-fp16"
train_data_dir = "C:/Users/Desktop/aase21/tst/1_girl"
output_dir = "E:/Apps/ComfyUI_windows_portable/ComfyUI/models/loras"
network_module = "networks.lora_lumina"
resolution = "1024,1024"
learning_rate = 1e-4
max_train_epochs = 1
#network_train_unet_only = true
ae = "E:/Apps/ComfyUI_windows_portable/ComfyUI/models/vae/ae.safetensors"
mixed_precision = "bf16"
gradient_checkpointing = true
cache_latents = true
cache_latents_to_disk = true
xformers = true
max_data_loader_n_workers = 1 
network_dim = 16
network_alpha = 8.0
optimizer_type = "AdamW8bit"
lr_scheduler = "cosine_with_restarts"
lr_warmup_steps = 10
lr_scheduler_num_cycles = 2
max_train_steps = 100
loss_type = "l2"
model_prediction_type = "sigma_scaled"
discrete_flow_shift = 3.0
min_timestep = 0
max_timestep = 1000
output_name = "lora_test" 
save_precision = "bf16"
save_model_as = "safetensors"
save_every_n_steps = 100
save_state = false
logging_dir = "E:/Apps/ComfyUI_windows_portable/ComfyUI/models/loras/logs"```

Would appreciate your insights on this! Thank you

**Comment by rockerBOO** [comment]
@GumGum10  One thing is the discrete_flow_shift should be like 6.0 for Lumina. Probably not causing the issue. Secondly maybe training the text encoder could be causing problems but I did not find any problems with my run. 

Also I didn't test AdamW8Bit but that might be a factor too. Xformers I didn't test but not sure if it was implemented for Lumina (I'd have to check) . model_prediction_type = "sigma_scaled" could be a factor. maybe `model_prediction_type = "raw"` would limit what it's doing. 

If you could try changing some of those and see if it is better/worse or the same. I can test this in a little bit.

**Comment by rockerBOO** [comment]
Example minimal config. Things commented out are supported but optional.

```toml

# bf16: https://huggingface.co/Comfy-Org/Lumina_Image_2.0_Repackaged/blob/main/split_files/diffusion_models/lumina_2_model_bf16.safetensors
# full precision: https://huggingface.co/rockerBOO/lumina-image-2/blob/main/lumina-image-2.safetensors
pretrained_model_name_or_path="lumina-image-2.safetensors" 

# https://huggingface.co/Comfy-Org/Lumina_Image_2.0_Repackaged/blob/main/split_files/text_encoders/gemma_2_2b_fp16.safetensors
gemma2="gemma-2-2b.safetensors" # Gemma 2 2B 

# https://huggingface.co/Comfy-Org/Lumina_Image_2.0_Repackaged/blob/main/split_files/vae/ae.safetensors
ae="ae.safetensors" # Same VAE as flux, 


# System prompt
# These are the 2 suggested. First is general purpose, where second is aligned with image-text, so good quality captioning. The system prompt is not super rigid, so you can play with these but these are what they may have pre-trained with

system_prompt = "You are an assistant designed to generate high-quality images based on user prompts."
# system_prompt = "You are an assistant designed to generate high-quality images with the highest degree of image-text alignment based on textual prompts."

# use_flash_attn = true # Use Flash Attention, needs pip install flash-attn

timestep_sampling = "nextdit_shift"
discrete_flow_shift = 6.0 # Default: 6.0
model_prediction_type = "raw" 
guidance_scale = 4.0

# cache_latents = true
# cache_latents_to_disk = true
# cache_text_encoder_outputs = true
# cache_text_encoder_outputs_to_disk = true

mixed_precision = "bf16"
# fp8_base = true 

network_module = "networks.lora_lumina"

network_dim = 8
network_alpha = 8

learning_rate = 1e-4
optimizer_type = "AdamW"
```

Sample prompts have 2 new values: `-ct` CFG truncate and `-rc` Renorm CFG, which have defaults. Usually can keep the defaults
 
```
-ct 0.25 -rc 1.0
```

Network args

```toml
network_args = [
   "attn_dim=4", # attention dimension
   "mlp_dim=4", # MLP dimension
   "mod_dim=4", # modulation dimension
   "refiner_dim=4", # refiner blocks dimension
  # Embedders condition the inputs, x = image, t = timestep, cap = caption
  # x_embedder, t_embedder, cap_embedder 
   "embedder_dims=[4,4,4]" 
]
```

**Comment by rockerBOO** [comment]
@GumGum10 I tried this a bunch of different ways 

- with/without TE training
- fp8/bf16

otherwise using the settings like above. I couldn't get it to distort as much as you're seeing, and 400 and 800 steps similar results. 


| sample1 | sample2 |
|--------|--------|
| women-lumina-2025-07-07-154825-86f22eb5_e000001_02_20250707163809_42 | women-lumina-2025-07-07-143645-0faafbeb_e000001_02_20250707152737_42 |

**Comment by rockerBOO** [comment]
> model_prediction_type = "sigma_scaled"

This was the issue because Lumina 2 is "reverse" of the typical flow and the sigmas were opposite of the timesteps. Using raw would solve it for now but I am looking to get a fix for that soon.

**Comment by kohya-ss** [comment]
> Example minimal config. Things commented out are supported but optional.

Thank you for the sample config, that's a great help, I'll check it out too.

**Comment by sdbds** [comment]
> Hi, really appreciate the work that you are doing on this! Me and a few other people have tried using this to train a LoRA on the new Neta finetune of Lumina2 (https://huggingface.co/neta-art/NetaLumina_Alpha). We've all experienced the same issue, after training for a few steps (usually ~90), the LoRA seems to explode. The inference images come up with a lot of noise, like the model is collapsing. Example: image
> 
> Config:
> 
> ```
> gemma2 = "E:/Apps/trainLumi/models/gemma-2-2b-fp16"
> train_data_dir = "C:/Users/Desktop/aase21/tst/1_girl"
> output_dir = "E:/Apps/ComfyUI_windows_portable/ComfyUI/models/loras"
> network_module = "networks.lora_lumina"
> resolution = "1024,1024"
> learning_rate = 1e-4
> max_train_epochs = 1
> #network_train_unet_only = true
> ae = "E:/Apps/ComfyUI_windows_portable/ComfyUI/models/vae/ae.safetensors"
> mixed_precision = "bf16"
> gradient_checkpointing = true
> cache_latents = true
> cache_latents_to_disk = true
> xformers = true
> max_data_loader_n_workers = 1 
> network_dim = 16
> network_alpha = 8.0
> optimizer_type = "AdamW8bit"
> lr_scheduler = "cosine_with_restarts"
> lr_warmup_steps = 10
> lr_scheduler_num_cycles = 2
> max_train_steps = 100
> loss_type = "l2"
> model_prediction_type = "sigma_scaled"
> discrete_flow_shift = 3.0
> min_timestep = 0
> max_timestep = 1000
> output_name = "lora_test" 
> save_precision = "bf16"
> save_model_as = "safetensors"
> save_every_n_steps = 100
> save_state = false
> logging_dir = "E:/Apps/ComfyUI_windows_portable/ComfyUI/models/loras/logs"```
> 
> Would appreciate your insights on this! Thank you
> ```

Recommended to use flash-attn for training...

**Comment by rockerBOO** [comment]
Some docs revision:

> `--model_prediction_type=<choice>` – Model prediction processing method. Options: `raw`, `additive`, `sigma_scaled`. Default `sigma_scaled`. **Recommended: `raw`**

Is this accurate that it defaults to sigma_scaled?

> `--use_sage_attn` – Use Sage Attention. 

Sage attention needs the library `pip install sageattention` and it is *inference only*

> `--timestep_sampling=<choice>` – Timestep sampling method. Options: `sigma`, `uniform`, `sigmoid`, `shift`, `nextdit_shift`. Default `sigma`. **Recommended: `nextdit_shift`**

This may need to be confirmed, not sure sigma is the option? And should probably default to `shift` if anything. 

> --use_flash_attn \

In the examples, might be difficult as it can be hard to install it. Would have it be recommended but not in the example.

**Comment by kohya-ss** [comment]
Thanks for checking the file!

> > `--model_prediction_type=<choice>` – Model prediction processing method. Options: `raw`, `additive`, `sigma_scaled`. Default `sigma_scaled`. **Recommended: `raw`**
> 
> Is this accurate that it defaults to sigma_scaled?

The default is `sigma_scaled`. I think because this is copied from `flux_train_util.py`. I will make `raw`  to default.

> > `--use_sage_attn` – Use Sage Attention.
> 
> Sage attention needs the library `pip install sageattention` and it is _inference only_

That's right. This option exists, but I've chosen not to mention it in the documentation.



> > `--timestep_sampling=<choice>` – Timestep sampling method. Options: `sigma`, `uniform`, `sigmoid`, `shift`, `nextdit_shift`. Default `sigma`. **Recommended: `nextdit_shift`**
> 
> This may need to be confirmed, not sure sigma is the option? And should probably default to `shift` if anything.

In my quick test, `nextdit_shift` works, but I will check with sigma and shift as well. I agree that the default shift seems good.

> > --use_flash_attn \
> 
> In the examples, might be difficult as it can be hard to install it. Would have it be recommended but not in the example.

You're certainly right.

**Comment by kohya-ss** [comment]
The prompt option `--ct` for sample image generation has been changed to `--ctr` because it conflicts with an existing option in `gen_img.py`. Also, `--rc` has been changed to `--rcfg` to avoid future conflicts. Thank you for your understanding.

**Comment by kohya-ss** [comment]
`lumina_train.py` seems to work now.

**Comment by rockerBOO** [comment]
There is a bug with the ModelOffloader for Flux in this PR. Passing the incorrect parameters into ModelOffloader(). I will post a fix soon.

**Comment by rockerBOO** [comment]
Hmm false alarm as it doesn't seem to have the issue in this PR. To confirm in flux_models.py https://github.com/kohya-ss/sd-scripts/pull/2138/files#diff-53ce178cab63cb26a2ef726379c796d907efe65a4e05c1552085a966a3f83fca removed the `self.num_double_blocks`. Flux training is working with this PR as far as I have tested.

**Comment by kohya-ss** [comment]
> Hmm false alarm as it doesn't seem to have the issue in this PR. To confirm in flux_models.py https://github.com/kohya-ss/sd-scripts/pull/2138/files#diff-53ce178cab63cb26a2ef726379c796d907efe65a4e05c1552085a966a3f83fca removed the `self.num_double_blocks`. Flux training is working with this PR as far as I have tested.

Thank you for confirming. I have confirmed that training with block swap enabled for Lumina/FLUX.1/SD3 works fine. I would like to merge this PR.

**Comment by kohya-ss** [comment]
The PR has finally been merged. Many thanks to all the contributions.
